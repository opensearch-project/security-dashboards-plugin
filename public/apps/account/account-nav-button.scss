/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

// stylelint-disable-next-line @osd/stylelint/no_modifying_global_selectors
.euiButtonEmpty.accountNavButton {
  border: 0;
}
