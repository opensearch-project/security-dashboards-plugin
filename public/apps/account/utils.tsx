/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { HttpStart } from 'opensearch-dashboards/public';
import { API_AUTH_LOGOUT, OPENID_AUTH_LOGOUT, SAML_AUTH_LOGOUT } from '../../../common';
import { API_ENDPOINT_ACCOUNT_INFO } from './constants';
import { AccountInfo } from './types';
import { httpGet, httpGetWithIgnores, httpPost } from '../configuration/utils/request-utils';
import { setShouldShowTenantPopup } from '../../utils/storage-utils';

export function fetchAccountInfo(http: HttpStart): Promise<AccountInfo> {
  return httpGet(http, API_ENDPOINT_ACCOUNT_INFO);
}

export async function fetchAccountInfoSafe(http: HttpStart): Promise<AccountInfo | undefined> {
  return httpGetWithIgnores<AccountInfo>(http, API_ENDPOINT_ACCOUNT_INFO, [401]);
}

export async function logout(http: HttpStart, logoutUrl?: string): Promise<void> {
  await httpPost(http, API_AUTH_LOGOUT);
  setShouldShowTenantPopup(null);
  // Clear everything in the sessionStorage since they can contain sensitive information
  sessionStorage.clear();
  // When no basepath is set, we can take '/' as the basepath.
  const basePath = http.basePath.serverBasePath ? http.basePath.serverBasePath : '/';
  const nextUrl = encodeURIComponent(basePath);
  window.location.href =
    logoutUrl || `${http.basePath.serverBasePath}/app/login?nextUrl=${nextUrl}`;
}

export async function samlLogout(http: HttpStart): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  window.location.href = `${http.basePath.serverBasePath}${SAML_AUTH_LOGOUT}`;
}

export async function openidLogout(http: HttpStart): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  sessionStorage.clear();
  window.location.href = `${http.basePath.serverBasePath}${OPENID_AUTH_LOGOUT}`;
}

export async function externalLogout(http: HttpStart, logoutEndpoint: string): Promise<void> {
  // This will ensure tenancy is picked up from local storage in the next login.
  setShouldShowTenantPopup(null);
  sessionStorage.clear();
  window.location.href = `${http.basePath.serverBasePath}${logoutEndpoint}`;
}

export async function updateNewPassword(
  http: HttpStart,
  newPassword: string,
  currentPassword: string
): Promise<void> {
  await httpPost(http, API_ENDPOINT_ACCOUNT_INFO, {
    password: newPassword,
    current_password: currentPassword,
  });
}
