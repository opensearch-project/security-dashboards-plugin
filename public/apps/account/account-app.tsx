/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import ReactDOM from 'react-dom';
import { CoreStart } from 'opensearch-dashboards/public';
import { AccountNavButton } from './account-nav-button';
import { fetchAccountInfoSafe } from './utils';
import { ClientConfigType } from '../../types';
import { AuthType, CUSTOM_ERROR_PAGE_URI, ERROR_MISSING_ROLE_PATH } from '../../../common';
import { fetchCurrentTenant, selectTenant } from '../configuration/utils/tenant-utils';
import {
  getSavedTenant,
  getShouldShowTenantPopup,
  setShouldShowTenantPopup,
} from '../../utils/storage-utils';
import { constructErrorMessageAndLog } from '../error-utils';
import { fetchCurrentAuthType } from '../../utils/logout-utils';
import { getDashboardsInfoSafe } from '../../utils/dashboards-info-utils';

function tenantSpecifiedInUrl() {
  return (
    window.location.search.includes('security_tenant') ||
    window.location.search.includes('securitytenant')
  );
}

export async function setupTopNavButton(coreStart: CoreStart, config: ClientConfigType) {
  const authType = config.auth?.type;
  let currAuthType = '';
  if (typeof authType === 'string') {
    currAuthType = authType;
  } else if (Array.isArray(authType) && authType.length === 1) {
    currAuthType = authType[0];
  } else {
    try {
      currAuthType = (await fetchCurrentAuthType(coreStart.http))?.currentAuthType;
    } catch (e) {
      currAuthType = AuthType.BASIC;
    }
  }

  const accountInfo = (await fetchAccountInfoSafe(coreStart.http))?.data;
  if (accountInfo) {
    // Missing role error
    if (accountInfo.roles.length === 0 && !window.location.href.includes(CUSTOM_ERROR_PAGE_URI)) {
      window.location.href =
        coreStart.http.basePath.serverBasePath + CUSTOM_ERROR_PAGE_URI + ERROR_MISSING_ROLE_PATH;
    }

    let tenant: string | undefined;
    if (config.multitenancy.enabled) {
      try {
        tenant = await fetchCurrentTenant(coreStart.http);
      } catch (e) {
        tenant = undefined;
        console.log(e);
      }
    }

    let shouldShowTenantPopup = true;

    const dashboardsInfo = await getDashboardsInfoSafe(coreStart.http);
    const multitenancyEnabled = dashboardsInfo?.multitenancy_enabled && config.multitenancy.enabled;
    const defaultTenantSet = dashboardsInfo?.default_tenant !== '';
    if (
      tenantSpecifiedInUrl() ||
      getShouldShowTenantPopup() === false ||
      !multitenancyEnabled ||
      defaultTenantSet
    ) {
      shouldShowTenantPopup = false;
    } else {
      // Switch to previous tenant based on localStorage, it may fail due to
      // 1) Localstorage is disabled; 2) Request failed
      try {
        const savedTenant = getSavedTenant();
        if (savedTenant !== null) {
          if (savedTenant === tenant) {
            shouldShowTenantPopup = false;
          } else {
            await selectTenant(coreStart.http, {
              username: accountInfo.user_name,
              tenant: savedTenant,
            });
            tenant = savedTenant;
            shouldShowTenantPopup = false;
            window.location.reload();
          }
        }
      } catch (e: any) {
        constructErrorMessageAndLog(e, `Failed to switch to ${tenant} tenant.`);
      }
    }

    setShouldShowTenantPopup(shouldShowTenantPopup);

    coreStart.chrome.navControls.registerRight({
      // Pin to rightmost, since newsfeed plugin is using 1000, here needs a number > 1000
      order: 2000,
      mount: (element: HTMLElement) => {
        ReactDOM.render(
          <AccountNavButton
            coreStart={coreStart}
            isInternalUser={accountInfo.is_internal_user}
            username={accountInfo.user_name}
            tenant={tenant}
            config={config}
            currAuthType={currAuthType.toLowerCase()}
          />,
          element
        );
        return () => ReactDOM.unmountComponentAtNode(element);
      },
    });
  }
}
