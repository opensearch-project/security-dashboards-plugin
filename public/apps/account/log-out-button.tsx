/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { EuiButtonEmpty } from '@elastic/eui';
import { HttpStart } from 'opensearch-dashboards/public';
import { externalLogout, logout } from './utils';
import { AuthType, OPENID_AUTH_LOGOUT, SAML_AUTH_LOGOUT } from '../../../common';
import { setShouldShowTenantPopup } from '../../utils/storage-utils';

export function LogoutButton(props: {
  authType: string;
  http: HttpStart;
  divider: JSX.Element;
  logoutUrl?: string;
}) {
  if (props.authType === AuthType.OPEN_ID) {
    return (
      <div>
        {props.divider}
        <EuiButtonEmpty
          data-test-subj="log-out-2"
          color="danger"
          size="xs"
          onClick={() => externalLogout(props.http, OPENID_AUTH_LOGOUT)}
        >
          Log out
        </EuiButtonEmpty>
      </div>
    );
  } else if (props.authType === AuthType.SAML) {
    return (
      <div>
        {props.divider}
        <EuiButtonEmpty
          data-test-subj="log-out-1"
          color="danger"
          size="xs"
          onClick={() => externalLogout(props.http, SAML_AUTH_LOGOUT)}
        >
          Log out
        </EuiButtonEmpty>
      </div>
    );
  } else if (props.authType === AuthType.PROXY) {
    setShouldShowTenantPopup(null);
    return <div />;
  } else {
    return (
      <div>
        {props.divider}
        <EuiButtonEmpty
          data-test-subj="log-out-3"
          color="danger"
          size="xs"
          onClick={() => logout(props.http, props.logoutUrl)}
        >
          Log out
        </EuiButtonEmpty>
      </div>
    );
  }
}
