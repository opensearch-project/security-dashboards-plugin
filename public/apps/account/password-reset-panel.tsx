/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import {
  EuiButton,
  EuiButtonEmpty,
  EuiCallOut,
  EuiFieldPassword,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFormRow,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiOverlayMask,
  EuiSpacer,
  EuiTitle,
} from '@elastic/eui';
import { CoreStart } from 'opensearch-dashboards/public';
import { FormRow } from '../configuration/utils/form-row';
import { logout, updateNewPassword } from './utils';
import { PASSWORD_INSTRUCTION } from '../apps-constants';
import { constructErrorMessageAndLog } from '../error-utils';
import { validateCurrentPassword } from '../../utils/login-utils';
import { getDashboardsInfo } from '../../utils/dashboards-info-utils';
import { PasswordStrengthBar } from '../configuration/utils/password-strength-bar';

interface PasswordResetPanelProps {
  coreStart: CoreStart;
  username: string;
  handleClose: () => void;
  logoutUrl?: string;
}

export function PasswordResetPanel(props: PasswordResetPanelProps) {
  const [currentPassword, setCurrentPassword] = React.useState<string>('');
  // reply on backend response of login call to verify
  const [isCurrentPasswordInvalid, setIsCurrentPasswordInvalid] = React.useState<boolean>(false);
  const [currentPasswordError, setCurrentPasswordError] = React.useState<string[]>([]);

  const [newPassword, setNewPassword] = React.useState<string>('');
  // reply on backend response of user update call to verify
  const [isNewPasswordInvalid, setIsNewPasswordInvalid] = React.useState<boolean>(false);

  const [repeatNewPassword, setRepeatNewPassword] = React.useState<string>('');
  const [isRepeatNewPasswordInvalid, setIsRepeatNewPasswordInvalid] = React.useState<boolean>(
    false
  );

  const [errorCallOut, setErrorCallOut] = React.useState<string>('');

  const [passwordHelpText, setPasswordHelpText] = React.useState<string>(PASSWORD_INSTRUCTION);

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        setPasswordHelpText(
          (await getDashboardsInfo(props.coreStart.http)).password_validation_error_message
        );
      } catch (e) {
        console.error(e);
      }
    };

    fetchData();
  }, [props.coreStart.http]);

  const handleReset = async () => {
    const http = props.coreStart.http;
    // validate the current password
    try {
      await validateCurrentPassword(http, props.username, currentPassword);
    } catch (e) {
      setIsCurrentPasswordInvalid(true);
      setCurrentPasswordError([constructErrorMessageAndLog(e, 'Invalid current password.')]);
    }

    // update new password
    try {
      await updateNewPassword(http, newPassword, currentPassword);

      await logout(http, props.logoutUrl);
    } catch (e) {
      setErrorCallOut(constructErrorMessageAndLog(e, 'Failed to reset password.'));
    }
  };

  // TODO: replace the instruction message for new password once UX provides it.
  return (
    <EuiOverlayMask>
      <EuiModal data-test-subj="reset-password-modal" onClose={props.handleClose}>
        <EuiSpacer />
        <EuiModalBody>
          <EuiTitle>
            <h4>Reset password for &quot;{props.username}&quot;</h4>
          </EuiTitle>

          <EuiSpacer />

          <FormRow
            headerText="Current password"
            helpText="Verify your account by entering your current password."
            isInvalid={isCurrentPasswordInvalid}
            error={currentPasswordError}
          >
            <EuiFieldPassword
              data-test-subj="current-password"
              onChange={function (e: React.ChangeEvent<HTMLInputElement>) {
                setCurrentPassword(e.target.value);
                setIsCurrentPasswordInvalid(false);
              }}
              isInvalid={isCurrentPasswordInvalid}
            />
          </FormRow>
          <EuiSpacer />

          <EuiFlexGroup direction="row">
            <EuiFlexItem grow={false}>
              <FormRow
                headerText="New password"
                helpText={passwordHelpText}
                isInvalid={isNewPasswordInvalid}
              >
                <EuiFieldPassword
                  data-test-subj="new-password"
                  onChange={function (e: React.ChangeEvent<HTMLInputElement>) {
                    setNewPassword(e.target.value);
                    setIsNewPasswordInvalid(false);
                    setIsRepeatNewPasswordInvalid(repeatNewPassword !== newPassword);
                  }}
                  isInvalid={isNewPasswordInvalid}
                />
              </FormRow>
              <EuiFormRow>
                <PasswordStrengthBar password={newPassword} />
              </EuiFormRow>
            </EuiFlexItem>
          </EuiFlexGroup>

          <FormRow
            headerText="Re-enter new password"
            helpText="The password must be identical to what you entered above."
          >
            <EuiFieldPassword
              data-test-subj="reenter-new-password"
              isInvalid={isRepeatNewPasswordInvalid}
              onChange={function (e: React.ChangeEvent<HTMLInputElement>) {
                const value = e.target.value;
                setRepeatNewPassword(value);
                setIsRepeatNewPasswordInvalid(value !== newPassword);
              }}
            />
          </FormRow>

          <EuiSpacer />

          {errorCallOut && (
            <EuiCallOut color="danger" iconType="alert">
              {errorCallOut}
            </EuiCallOut>
          )}
        </EuiModalBody>
        <EuiModalFooter>
          <EuiButtonEmpty data-test-subj="cancel" onClick={props.handleClose}>
            Cancel
          </EuiButtonEmpty>

          <EuiButton
            data-test-subj="reset"
            fill
            disabled={isRepeatNewPasswordInvalid}
            onClick={handleReset}
          >
            Reset
          </EuiButton>
        </EuiModalFooter>
      </EuiModal>
    </EuiOverlayMask>
  );
}
