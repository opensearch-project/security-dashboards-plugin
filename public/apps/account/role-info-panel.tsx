/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */
import {
  EuiModal,
  EuiModalBody,
  EuiOverlayMask,
  EuiText,
  EuiTitle,
  EuiHorizontalRule,
  EuiSpacer,
} from '@elastic/eui';
import React, { useCallback } from 'react';
import { CoreStart } from 'opensearch-dashboards/public';
import { fetchAccountInfo } from './utils';

export function RoleInfoPanel(props: { coreStart: CoreStart; handleClose: () => void }) {
  const [roles, setRoles] = React.useState<string[]>([]);
  const [backendRoles, setBackendRoles] = React.useState<string[]>([]);

  const fetchData = useCallback(async () => {
    try {
      const accountInfo = await fetchAccountInfo(props.coreStart.http);
      setRoles(accountInfo?.data.roles || []);
      setBackendRoles(accountInfo?.data.backend_roles || []);
    } catch (e) {
      console.log(e);
    }
  }, [props.coreStart.http]);

  React.useEffect(() => {
    fetchData();
  }, [props.coreStart.http, fetchData]);

  return (
    <EuiOverlayMask>
      <EuiModal data-test-subj="role-info-modal" onClose={props.handleClose}>
        <EuiSpacer />
        <EuiModalBody>
          <EuiTitle>
            <h4>Roles ({roles.length})</h4>
          </EuiTitle>
          <EuiText color="subdued">
            Roles you are currently mapped to by your administrator.
          </EuiText>
          <EuiSpacer />
          {roles.map((item) => (
            <EuiText key={item}>
              {item}
              <br />
            </EuiText>
          ))}
          <EuiHorizontalRule />
          <EuiTitle>
            <h4>Backend roles ({backendRoles.length})</h4>
          </EuiTitle>
          <EuiText color="subdued">
            Backend roles you are currently mapped to by your administrator.
          </EuiText>
          <EuiSpacer />
          {backendRoles.map((item) => (
            <EuiText key={item}>
              {item}
              <br />
            </EuiText>
          ))}
        </EuiModalBody>
      </EuiModal>
    </EuiOverlayMask>
  );
}
