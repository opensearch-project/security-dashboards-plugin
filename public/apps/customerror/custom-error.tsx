/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiButton, EuiImage, EuiListGroup, EuiSpacer, EuiText } from '@elastic/eui';
import { AppMountParameters, CoreStart } from 'opensearch-dashboards/public';
import React from 'react';
import ReactDOM from 'react-dom';
import { Router, Route } from 'react-router-dom';
import { ERROR_MISSING_ROLE_PATH } from '../../../common';
import { ClientConfigType } from '../../types';
import { AuthType } from '../../../common';
import './_index.scss';
import { logout } from '../account/utils';

interface CustomErrorDeps {
  title: string;
  subtitle: string;
  http: CoreStart['http'];
  chrome: CoreStart['chrome'];
  config: ClientConfigType['ui'][AuthType.BASIC]['login'];
}

export function CustomErrorPage(props: CustomErrorDeps) {
  return (
    <EuiListGroup className="custom-error-wrapper">
      {props.config.showbrandimage && (
        <EuiImage alt="" url={props.config.brandimage || props.chrome.logos.OpenSearch.url} />
      )}
      <EuiSpacer size="s" />
      <EuiText size="m" textAlign="center">
        <h3>{props.title}</h3>
      </EuiText>
      <EuiSpacer size="s" />
      <EuiText size="s" textAlign="center">
        {props.subtitle}
      </EuiText>
      <EuiSpacer size="s" />
      <EuiButton
        fill
        onClick={() => logout(props.http, '')}
        data-test-subj="error-logout-button"
        fullWidth
      >
        Logout
      </EuiButton>
    </EuiListGroup>
  );
}

export async function renderPage(
  coreStart: CoreStart,
  params: AppMountParameters,
  config: ClientConfigType
) {
  ReactDOM.render(
    <Router history={params.history}>
      <Route path={ERROR_MISSING_ROLE_PATH}>
        <CustomErrorPage
          http={coreStart.http}
          chrome={coreStart.chrome}
          config={config.ui.basicauth.login}
          title="Missing Role"
          subtitle="No roles available for this user, please contact your system administrator."
        />
        ,
      </Route>
    </Router>,
    params.element
  );
  return () => ReactDOM.unmountComponentAtNode(params.element);
}
export { EuiButton };
