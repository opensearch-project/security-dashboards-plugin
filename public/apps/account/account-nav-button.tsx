/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import {
  EuiAvatar,
  EuiButtonEmpty,
  EuiContextMenuPanel,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHeaderSectionItemButton,
  EuiHorizontalRule,
  EuiListGroup,
  EuiListGroupItem,
  EuiPopover,
  EuiText,
} from '@elastic/eui';
import { CoreStart } from 'opensearch-dashboards/public';
import React, { useCallback } from 'react';
import { RoleInfoPanel } from './role-info-panel';
import { PasswordResetPanel } from './password-reset-panel';
import { TenantSwitchPanel } from './tenant-switch-panel';
import { ClientConfigType } from '../../types';
import { LogoutButton } from './log-out-button';
import { resolveTenantName } from '../configuration/utils/tenant-utils';
import { getShouldShowTenantPopup, setShouldShowTenantPopup } from '../../utils/storage-utils';
import { getDashboardsInfo } from '../../utils/dashboards-info-utils';

export function AccountNavButton(props: {
  coreStart: CoreStart;
  isInternalUser: boolean;
  username: string;
  tenant?: string;
  config: ClientConfigType;
  currAuthType: string;
}) {
  const [isPopoverOpen, setPopoverOpen] = React.useState<boolean>(false);
  const [modal, setModal] = React.useState<React.ReactNode>(null);
  const horizontalRule = <EuiHorizontalRule margin="xs" />;
  const username = props.username;
  const [isMultiTenancyEnabled, setIsMultiTenancyEnabled] = React.useState<boolean>(true);

  const showTenantSwitchPanel = useCallback(
    () =>
      setModal(
        <TenantSwitchPanel
          coreStart={props.coreStart}
          config={props.config}
          handleClose={() => {
            setModal(null);
          }}
          handleSwitchAndClose={() => {
            setModal(null);
            reloadAfterTenantSwitch();
          }}
          tenant={props.tenant!}
        />
      ),
    [props.config, props.coreStart, props.tenant]
  );
  React.useEffect(() => {
    const fetchData = async () => {
      try {
        setIsMultiTenancyEnabled(
          (await getDashboardsInfo(props.coreStart.http)).multitenancy_enabled
        );
      } catch (e) {
        // TODO: switch to better error display.
        console.error(e);
      }
    };

    fetchData();
  }, [props.coreStart.http]);

  // Check if the tenant modal should be shown on load
  if (getShouldShowTenantPopup()) {
    setShouldShowTenantPopup(false);
    showTenantSwitchPanel();
  }

  const contextMenuPanel = (
    <div style={{ maxWidth: '256px' }}>
      <EuiFlexGroup gutterSize="s">
        <EuiFlexItem grow={null}>
          <EuiAvatar name={username} />
        </EuiFlexItem>
        <EuiFlexItem>
          <EuiListGroup gutterSize="none">
            <EuiListGroupItem
              key="username"
              wrapText
              label={
                <EuiText size="s">
                  <h5>{username}</h5>
                </EuiText>
              }
            />
          </EuiListGroup>
          {isMultiTenancyEnabled && props.config.multitenancy.enabled && (
            <EuiListGroupItem
              color="subdued"
              key="tenant"
              label={
                <EuiText size="xs" id="tenantName">
                  {resolveTenantName(props.tenant || '', username)}
                </EuiText>
              }
            />
          )}
        </EuiFlexItem>
      </EuiFlexGroup>

      {horizontalRule}
      <EuiButtonEmpty
        data-test-subj="view-roles-and-identities"
        size="xs"
        onClick={() => setModal(<RoleInfoPanel {...props} handleClose={() => setModal(null)} />)}
      >
        View roles and identities
      </EuiButtonEmpty>
      {isMultiTenancyEnabled && props.config.multitenancy.enabled && (
        <>
          {horizontalRule}
          <EuiButtonEmpty data-test-subj="switch-tenants" size="xs" onClick={showTenantSwitchPanel}>
            Switch tenants
          </EuiButtonEmpty>
        </>
      )}
      {props.isInternalUser && (
        <>
          {horizontalRule}
          <EuiButtonEmpty
            data-test-subj="reset-password"
            size="xs"
            onClick={() =>
              setModal(
                <PasswordResetPanel
                  coreStart={props.coreStart}
                  username={props.username}
                  handleClose={() => setModal(null)}
                  logoutUrl={props.config.auth.logout_url}
                />
              )
            }
          >
            Reset password
          </EuiButtonEmpty>
        </>
      )}
      <LogoutButton
        authType={props.currAuthType}
        http={props.coreStart.http}
        divider={horizontalRule}
        logoutUrl={props.config.auth.logout_url}
      />
    </div>
  );
  return (
    <EuiHeaderSectionItemButton id="user-icon-btn">
      <EuiPopover
        data-test-subj="account-popover"
        id="actionsMenu"
        button={<EuiAvatar name={username} />}
        isOpen={isPopoverOpen}
        closePopover={() => {
          setPopoverOpen(false);
        }}
        panelPaddingSize="s"
        onClick={() => {
          setPopoverOpen((prevState) => !prevState);
        }}
      >
        <EuiContextMenuPanel>{contextMenuPanel}</EuiContextMenuPanel>
      </EuiPopover>
      {modal}
    </EuiHeaderSectionItemButton>
  );
}

export function reloadAfterTenantSwitch(): void {
  // the below portion is to clear URLs starting with 'lastUrl'
  // when switching tenants, the last URLs will be from the old tenancy therefore we need to remove these from sessionStorage.
  const lastUrls = [];
  const { sessionStorage } = window;

  for (let i = 0; i < sessionStorage.length; i++) {
    const key = sessionStorage.key(i);
    if (key?.startsWith('lastUrl')) {
      lastUrls.push(key);
    }
  }
  for (let i = 0; i < lastUrls.length; i++) {
    sessionStorage.removeItem(lastUrls[i]);
  }

  // rather than just reload when we switch tenants, we set the URL to the pathname. i.e. the portion like: '/app/dashboards'
  // therefore, the copied URL will now allow tenancy changes.
  window.location.href = window.location.pathname;
}
