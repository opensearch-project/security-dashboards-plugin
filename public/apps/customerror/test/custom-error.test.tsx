/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */
import { shallow } from 'enzyme';
import React from 'react';
import { CustomErrorPage } from '../custom-error';
import { cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import { EuiButton } from '../custom-error';
import { logout } from '../../account/utils';

afterEach(() => {
  cleanup();
});

describe('Custom error page test', () => {
  let component;

  beforeEach(() => {
    component = shallow(
      <EuiButton fill onClick={logout} data-test-subj="error-logout-button" fullWidth>
        Logout
      </EuiButton>
    );
  });

  it('renders and clicks the button on the error page', () => {
    const wrapper = shallow(
      <CustomErrorPage
        title="Title"
        subtitle="Sub Title"
        http={undefined}
        chrome={undefined}
        config={{
          title: '',
          subtitle: '',
          showbrandimage: false,
          brandimage: '',
          buttonstyle: '',
        }}
      />
    );

    expect(wrapper).toMatchSnapshot();

    component.find('[data-test-subj="error-logout-button"]').simulate('onClick', {
      preventDefault: () => {},
    });
  });
});
