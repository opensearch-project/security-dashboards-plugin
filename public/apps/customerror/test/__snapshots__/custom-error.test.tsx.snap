// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom error page test renders and clicks the button on the error page 1`] = `
<EuiListGroup
  className="custom-error-wrapper"
>
  <EuiSpacer
    size="s"
  />
  <EuiText
    size="m"
    textAlign="center"
  >
    <h3>
      Title
    </h3>
  </EuiText>
  <EuiSpacer
    size="s"
  />
  <EuiText
    size="s"
    textAlign="center"
  >
    Sub Title
  </EuiText>
  <EuiSpacer
    size="s"
  />
  <EuiButton
    data-test-subj="error-logout-button"
    fill={true}
    fullWidth={true}
    onClick={[Function]}
  >
    Logout
  </EuiButton>
</EuiListGroup>
`;
